Sources/xcodeproj/Utils/PBXBatchUpdater.swift

```diff
@@ -0,0 +1,83 @@
+import Foundation
+import PathKit
+
+// This is a helper class for quickly adding a large number of files.
+// It is forbidden to add a file to a group one by one using the PBXGroup method addFile(...) while you are working with this class.
+public final class PBXBatchUpdater {
+    private let projectObjects: PBXObjects
+    private let sourceRoot: Path
+    private var references: [Path: PBXObjectReference]?
+
+    init(projectObjects: PBXObjects, sourceRoot: Path) {
+        self.projectObjects = projectObjects
+        self.sourceRoot = sourceRoot
+    }
+
+    /// Adds file at the give path to the project or returns existing file and its reference.
+    ///
+    /// - Parameters:
+    ///   - fileName: file name.
+    ///   - sourceTree: file sourceTree, default is `.group`
+    ///   - sourceRoot: path to project's source root.
+    /// - Returns: new or existing file and its reference.
+    @discardableResult
+    public func addFile(
+        to group: PBXGroup,
+        fileName: String,
+        sourceTree: PBXSourceTree = .group
+    ) throws -> PBXFileReference {
+        let groupPath = try group.fullPath(sourceRoot: sourceRoot)
+        let filePath = (groupPath ?? sourceRoot) + Path(fileName)
```

Comment:
I think this fallback to `sourceRoot` is not obvious. I'd be strict and assume that `fullPath` returns a path. If it doesn't, that probably indicates that the project has been misconfigured and thus, it's something to be looked at at their end.

Reply:
I agree, so I do not understand why the `fullPath` method returns the optional value.